components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/DefaultServiceURLBuilder.java

```diff
@@ -0,0 +1,212 @@
+
+import java.net.SocketException;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.core.util.IdentityTenantUtil.isTenantQualifiedUrlsEnabled;
+
+/**
+ * Implementation for {@link ServiceURLBuilder}.
+ * Builder for {@link ServiceURL} instances.
+ */
+public class DefaultServiceURLBuilder implements ServiceURLBuilder {
+
+    private Boolean addProxyContext;
+    private Boolean addWebContext;
+    private String urlContext;
+    private Boolean appendTenantAsQueryParam;
+    private Boolean appendTenantAsPathParam;
+    private Map<String, Object> parameters;
+    private Map<String, Object> fragments;
+
+    @Override
+    public ServiceURLBuilder addContext(String context) {
+
+        this.urlContext = context;
+        return this;
+    }
+
+    @Override
+    public final ServiceURLBuilder addProxyContext() {
+
+        this.addProxyContext = true;
+        return this;
+    }
+
+    @Override
+    public final ServiceURLBuilder addWebContext() {
+
+        this.addWebContext = true;
+        return this;
+    }
+
+    @Override
+    public ServiceURLBuilder appendTenantAsQueryParam() {
+
+        this.appendTenantAsQueryParam = true;
+        return this;
+    }
+
+    @Override
+    public ServiceURLBuilder appendTenantAsPathParam() {
+
+        this.appendTenantAsPathParam = true;
+        return this;
+    }
+
+    @Override
+    public ServiceURLBuilder addParameter(String key, Object value) {
+
+        return null;
+    }
+
+    @Override
+    public ServiceURL build() throws URLResolverException {
+
+        String protocol = fetchProtocol();
+        String hostName = fetchHostName();
+        Integer port = fetchPort();
+        String proxyContext = null;
+        if (addProxyContext) {
+            proxyContext = fetchProxyContextPath();
+        }
+        String webContext = null;
+        if (addWebContext) {
+            webContext = fetchWebContextRoot();
+        }
+
+        String urlContext = this.urlContext;
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+        if (StringUtils.isBlank(tenantDomain)) {
+            tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        }
+
+        String resolvedUrlContext = null;
+        StringBuilder resolvedUrlStringBuilder = new StringBuilder();
+        if (isTenantQualifiedUrlsEnabled()) {
+            if (StringUtils.isNotBlank(tenantDomain)) {
+                resolvedUrlStringBuilder.append("/t/").append(tenantDomain);
+            }
+            resolvedUrlContext = resolveUrlFromStandardContext(urlContext);
+        }
+
+        if (!isTenantQualifiedUrlsEnabled() && appendTenantAsPathParam) {
+            if (StringUtils.isNotBlank(tenantDomain) &&
+                    !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equalsIgnoreCase(tenantDomain)) {
+                resolvedUrlStringBuilder.append("/t/").append(tenantDomain);
+            }
+        }
```

Comment:
If not tenant url supported we should give precedence to load the URL from the config